rust/sbp/src/link.rs

```diff
@@ -0,0 +1,369 @@
+    }
+}
+
+impl<'link, S> Clone for Link<'link, S> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+}
+
+struct LinkInner<'link, S> {
+    handlers: Mutex<DenseSlotMap<KeyInner, Handler<'link, S>>>,
+}
+
+pub struct Handler<'link, S> {
+    func: Box<dyn FnMut(&S, SBP) + Send + 'link>,
+    msg_types: Cow<'static, [u16]>,
+}
+
+impl<'link, S> Handler<'link, S> {
+    fn run(&mut self, state: &S, msg: SBP) {
+        (self.func)(state, msg);
+    }
+
+    fn can_run(&self, msg: &SBP) -> bool {
+        self.msg_types.contains(&msg.get_message_type()) || self.msg_types.is_empty()
+    }
+}
+
+pub trait IntoHandler<'link, S, E, HandlerKind = WithState> {
+    fn into_handler(self) -> Handler<'link, S>;
+    fn into_handler_with_ids(self, msg_types: &[u16]) -> Handler<'link, S>;
+}
+
+pub struct WithState;
+
+impl<'link, S, E, F> IntoHandler<'link, S, E, WithState> for F
+where
+    E: Event,
+    F: FnMut(&S, E) + Send + 'link,
+{
+    fn into_handler(mut self) -> Handler<'link, S> {
+        Handler {
+            func: Box::new(move |state, msg| {
+                let event = E::from_sbp(msg);
+                self(state, event);
+            }),
+            msg_types: Cow::Borrowed(E::MESSAGE_TYPES),
+        }
+    }
+
+    fn into_handler_with_ids(mut self, msg_types: &[u16]) -> Handler<'link, S> {
+        Handler {
+            func: Box::new(move |state, msg| {
+                let event = E::from_sbp(msg);
+                self(state, event);
+            }),
+            msg_types: Cow::Owned(msg_types.to_vec()),
+        }
+    }
+}
+
+pub struct WithoutState;
+
+impl<'link, S, E, F> IntoHandler<'link, S, E, WithoutState> for F
+where
+    E: Event,
+    F: FnMut(E) + Send + 'link,
+{
+    fn into_handler(mut self) -> Handler<'link, S> {
+        Handler {
+            func: Box::new(move |_state, msg| {
+                let event = E::from_sbp(msg);
+                self(event);
+            }),
+            msg_types: Cow::Borrowed(E::MESSAGE_TYPES),
+        }
+    }
+
+    fn into_handler_with_ids(mut self, msg_types: &[u16]) -> Handler<'link, S> {
+        Handler {
+            func: Box::new(move |_state, msg| {
+                let event = E::from_sbp(msg);
+                self(event);
+            }),
+            msg_types: Cow::Owned(msg_types.to_vec()),
+        }
+    }
+}
+
+slotmap::new_key_type! {
+    struct KeyInner;
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Key {
+    key: KeyInner,
+}
+
+pub trait Event {
+    const MESSAGE_TYPES: &'static [u16];
+
+    fn from_sbp(msg: SBP) -> Self;
+}
+
+impl Event for SBP {
+    const MESSAGE_TYPES: &'static [u16] = &[];
+
+    fn from_sbp(msg: SBP) -> Self {
+        msg
+    }
+}
+
+impl<T> Event for T
+where
+    T: ConcreteMessage,
+{
+    const MESSAGE_TYPES: &'static [u16] = &[T::MESSAGE_TYPE];
+
+    fn from_sbp(msg: SBP) -> Self {
+        msg.try_into().unwrap()
```

Comment:
Should this be fallible?  At least should use `.expect` unless we never expect `.unwrap` to fail?

Reply:
It shouldn't ever fail, because `from_sbp` is only called after checking that the message was in `MESSAGE_TYPES`. So if you create a custom event and set the wrong messages types you could hit this error. So an expect is probably a good idea to make that more clear

Reply:
Agree, expect could point out that it should only happen if there's a programming error

Reply:
Well so actually this shouldn't ever error, this blanket impl makes all `ConcreteMessage`s usable as events triggered by that messages type. So baring a logic error in the Link (i.e. sending an event to the wrong handler or something) this should never happen. Replaced the unwrap with a panic that will include the offending message just in case

Reply:
👍